api/v1alpha1/envoypatchpolicy_types.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gwapiv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+const (
+	// KindEnvoyPatchPolicy is the name of the EnvoyPatchPolicy kind.
+	KindEnvoyPatchPolicy = "EnvoyPatchPolicy"
+)
+
+// +kubebuilder:object:root=true
+
+// EnvoyPatchPolicy allows the user to modify the generated Envoy xDS
+// resources by Envoy Gateway using this patch API
+type EnvoyPatchPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of EnvoyPatchPolicy.
+	Spec EnvoyPatchPolicySpec `json:"spec"`
+
+	// Status defines the current status of EnvoyPatchPolicy.
+	Status EnvoyPatchPolicyStatus `json:"status"`
+}
+
+// EnvoyPatchPolicySpec defines the desired state of RateLimitFilter.
+// +union
+type EnvoyPatchPolicySpec struct {
+	// Type decides the type of patch.
+	// Valid EnvoyPatchType values are "JSONPatch".
+	//
+	// +unionDiscriminator
```

Comment:
this something like `oneOf` in protobuf?

Reply:
yup